### PR TITLE
[PHT] Complete node splitting

### PR DESCRIPTION
**Pull request by kaldoran**
This PR is finally the last update on #61.
I've tested it with benchmark and it works, could test it too with 2 dhtnode and work like a charm :D 

(Sadly this only add 300 lines for lots of time consumed by error and little fix in API and what so ever .)

### CODE REVIEW

**Comment by kaldoran** [comment]
GRRRRrrrrrrr ! Again this compilation error.

**Comment by sim590** [comment]
I think you need to rebase onto `index` branch.

**Comment by kaldoran** [comment]
Already done

**Comment by kaldoran** [comment]
And done ! 
[Also cherry-pick docker file onto index]

File: include/opendht/indexation/pht.h [hunk 1]
```diff
@@ -268,6 +344,137 @@ class Pht {
...
+            on_done,
+            pht_filter
+        );
+
+        dht_->get(sibling->hash(),
+            count,
+            on_done,
+            pht_filter
+        );
+    }
```
Comment (sim590) [inline include/opendht/indexation/pht.h]: This definition is too long to be in header. Move it in cpp.

File: include/opendht/indexation/pht.h [hunk 2]
```diff
@@ -159,7 +187,11 @@ class Pht {
      * serialization order of fields. */
     using KeySpec = std::map<std::string, size_t>;
 
+    /* Callback used for insert value by using the pht */
+    using RealInsertCallback = std::function<void(std::shared_ptr<Prefix> p, IndexEntry entry )>;
     using LookupCallback = std::function<void(std::vector<std::shared_ptr<Value>>& values, Prefix p)>;
+    using LookupCallbackWrapper = std::function<void(std::vector<std::shared_ptr<IndexEntry>>& values, Prefix p)>;
```
Comment (sim590) [inline include/opendht/indexation/pht.h]: Avoid exposing callbacks that won't be used in API functions.

File: include/opendht/indexation/pht.h [hunk 3]
```diff
@@ -268,6 +344,137 @@ class Pht {
...
+                }
+
+                return true;
+            },
+            [=](const dht::Value& v) {
+                /* Filter value v thats start with the same name as ours */
+                return v.user_type.compare(0, name_.size(), name_) == 0;
+            }
+        );
+    }
```
Comment (sim590) [inline include/opendht/indexation/pht.h]: Move in cpp

File: include/opendht/indexation/pht.h [hunk 4]
```diff
@@ -268,6 +344,137 @@ class Pht {
...
+        auto full = Prefix(entry.prefix);
+
+        auto loc = foundSplitLocation(full, vals);
+        auto prefix_to_insert = std::make_shared<Prefix>(full.getPrefix(loc));
+
+        for(;loc != insert.size_ - 1; loc--)
+            updateCanary(full.getPrefix(loc));
+
+        end_cb(prefix_to_insert, entry);
+    }
```
Comment (sim590) [inline include/opendht/indexation/pht.h]: Move in cpp too. IMO, if it makes more than around 5 lines, it should be defined in cpp as much as possible..

File: include/opendht/indexation/pht.h [hunk 5]
```diff
@@ -196,11 +228,42 @@ class Pht {
...
+     * @param kp          : Prefix to insert (linearize the the key)
+     * @param entry       : Entry created from the value
+     * @param lo          : Lowest point to start in the prefix
+     * @param hi          : Highest point to end in the prefix
+     * @param time_p      : Timepoint to use for the insertion into the dht (must be < now)
+     * @param check_split : If this flag is true then the algoritm will not use the merge algorithm
+     * @param done_cb     : Callback to call when the insert is done
+     */
+
+    void insert( Prefix kp, IndexEntry entry, std::shared_ptr<int> lo, std::shared_ptr<int> hi, time_point time_p,
```
Comment (sim590) [inline include/opendht/indexation/pht.h]: By the way, you could erase that unecessary white space between `(` and `Prefix` ;)

File: src/dht.cpp [hunk 6]
```diff
@@ -1245,6 +1243,9 @@ Dht::announce(const InfoHash& id, sa_family_t af, std::shared_ptr<Value> value,
             callback(false, {});
         return;
     }
+
+    storageStore(id, value, created);
+
```
Comment (sim590) [inline src/dht.cpp]: This change has been introduced by `bbd1200`. You should have cherry-picked it. Now, it won't be clear in the log if index has this fix or not.

File: tools/dhtnode.cpp [hunk 7]
```diff
@@ -284,7 +284,7 @@ void cmd_loop(std::shared_ptr<DhtRunner>& dht, std::map<std::string, indexation:
                 );
             }
             catch (std::invalid_argument& e) { std::cout << e.what() << std::endl; }
-            catch (std::out_of_range& e) { }
+            catch (std::out_of_range& e) { std::cout << e.what() << std::endl; }
```
Comment (sim590) [inline tools/dhtnode.cpp]: Is it useful to output this? Can you clarify?

**Comment by kaldoran** [comment]
Does that need more patch?

**Comment by kaldoran** [comment]
Sorry for this update, really though i made the needed change

**Pull request merged**